Adhere to BUIP 40 for max sigop consensus rules

jamoes opened the pull request:
[BUIP 40](https://github.com/BitcoinUnlimited/BUIP/blob/master/040.mediawiki) outlines Bitcoin Unlimited's approach to dealing with sigop counts in blocks greater than 1 MB. This change will modify Bitcoin Classic so that it uses the same approach as Bitcoin Unlimited's default values when dealing with sigops in large blocks. This change does not add any additional configuration options for "excessive transaction size" or "excessive sigops per MB".

Adds a new Policy::blockSigOpAcceptLimit function, which calculates the maximum allowed sigops for a given block size. This rounds the block size up to the nearest MB, and then allows 20,000 sigops per
MB of blockspace.

Renames MAX_BLOCK_SIGOPS to MAX_BLOCK_SIGOPS_PER_MB, and defines it directly, rather than as a factor of MAX_BLOCK_SIZE.

Adds unit tests for Policy::blockSigOpAcceptLimit.

Updates the QA test p2p-fullblocktest.py to test the new sigop rules. Additionally modifies this test to create 2 MB blocks, which is required to test the new sigop behavior.

Note that this does not modify the MAX_STANDARD_TX_SIGOPS constant or make any modifications to what is considered a standard transaction.

jamoes committed:
Adhere to BUIP 40 for max sigop consensus rules

Adds a new Policy::blockSigOpAcceptLimit function, which calculates
the maximum allowed sigops for a given block size. This rounds the
block size up to the nearest MB, and then allows 20,000 sigops per
MB of blockspace.

Renames MAX_BLOCK_SIGOPS to MAX_BLOCK_SIGOPS_PER_MB, and define it
directly, rather than as a factor of MAX_BLOCK_SIZE.

Adds unit tests for Policy::blockSigOpAcceptLimit.

Updates the QA test p2p-fullblocktest.py to test the new sigop
rules. Additionally modifies this test to create 2 MB blocks, which
is required to test the new sigop behavior.

zander commented on src/test/sigopcount_tests.cpp:
```diff
@@ -64,4 +65,16 @@ BOOST_AUTO_TEST_CASE(GetSigOpCount)
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(0), 20000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(100), 20000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(50000), 20000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(1e6), 20000);
```
What do you think about adding this test;

    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(1700000), 40000);

edit; added a zero.

jamoes replied:
Added!

zander commented on src/policy/policy.cpp:
```diff
@@ -183,3 +183,9 @@ int32_t Policy::blockSizeAcceptLimit()
+
+int64_t Policy::blockSigOpAcceptLimit(int32_t nBlockSize)
+{
+    uint64_t nBlockSizeMb = 1 + (nBlockSize - 1) / 1000000;
```
I expected something using std::ceil()

deadalnix replied:
No ceil works on floating point. You don't want floating point in consensus code, as they tend to be non deterministic. Notably on x86 rounding is controlled by a flag register that isn't saved/restored as part of the ABI, even when doing round trip to the kernel. There are war story about device driver changing the floating point rounding mode and causing application to misbehave.

For clarity, please put () around the division.

deadalnix replied:
No ceil. It is floating point. You don't want floating point in consensus code as it is non deterministic. If you want to be scared, know that the rounding mode on X86 is controlled by a flag register that isn't saved/restored by the ABI, even when doing round trips to the kernel, and there are war stories out there of driver changing the rounding mode during an IRQ.

zander replied:
Yes, I retract my suggestion for change. The current code works fine.

jamoes replied:
> For clarity, please put () around the division.

I added parens around the division for clarity.

zander reviewed (changes requested):
I like the setup and the changes, thanks!

deadalnix commented on src/policy/policy.h:
```diff
@@ -68,6 +68,7 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
 
 namespace Policy {
     std::int32_t blockSizeAcceptLimit();
+    std::int64_t blockSigOpAcceptLimit(int32_t nBlockSize);
```
This limits the block size to 4GB. Not sure this is a major concern at this point.

zander replied:
its an int, not unsigned, so only 2GB. :)
No, I don't think this is a concern any time soon.

deadalnix commented on src/test/sigopcount_tests.cpp:
```diff
@@ -64,4 +65,16 @@ BOOST_AUTO_TEST_CASE(GetSigOpCount)
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(1e6), 20000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(1e6 + 1), 40000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(2e6), 40000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(4e6 + 1), 100000);
```
It in unclear why these test case are chosen rather than others.

A few things. First 1e6 is a floating point, and you don't want to use that. This is annoying, but 10000000 is the way to go. Create a constant if that is too constraining. It's probably fine in that range, but test with larger value and you notice that 1eN + 1 == 1eN for larges values of N.

The exact value of N will depend of the rounding mode the CPU is in when running the code and the precision that the compiler chose to use, both of which are non standard.

I would test for a set of  X*1000000, X*1000000+1 and X*1000000-1 for various values of X, typical edges cases values (0, 1, uint.max)  and then some randomly picked values.

zquestz replied:
So when I see multiple checks like this, I think of table driven tests, they are far more clear, and easier to add additional cases. I generally do this with Golang not C++, not sure if that is something to strive for in the codebase. Here is an example in Golang: https://github.com/golang/go/wiki/TableDrivenTests

zander replied:
@zquestz this is a good point, and I miss the test library I've used for many years.

Boost has added this feature in 1.59 which is newer than we allow.

http://www.boost.org/doc/libs/1_63_0/libs/test/doc/html/boost_test/tests_organization/test_cases/test_case_generation.html

zquestz replied:
Gotcha, good to know. Any plan to try and upgrade the version of Boost used within the codebase?

jamoes replied:
I updated the test values to not use 1e6 notation. 

> I would test for a set of X1000000, X1000000+1 and X*1000000-1 for various values of X

I added these tests for X values of 1 and 2.

> typical edges cases values (0, 1, uint.max)

I added these tests as well.

> table driven tests

I looked into achieving this by simply using a std::map, and iterating over the map. I think in this case it made the test less legible (since it added lines of code and increased cognitive load on the reader), and made test failures harder to find (because the exact line number wasn't specified). I think this would make more sense when the function being tested is more complex, or if we had a better framework for doing data-driven tests.

zquestz replied:
Yeah agreed. Thanks for looking into it though!

deadalnix commented on src/miner.cpp:
```diff
@@ -251,9 +251,10 @@ CBlockTemplate* Mining::CreateNewBlock(const CChainParams& chainparams) const
             if (!IsFinalTx(tx, nHeight, nLockTimeCutoff))
                 continue;
 
+            const uint64_t maxSigOps = Policy::blockSigOpAcceptLimit(nBlockSize + nTxSize);
```
This is absolutely huge and may win being rearchitectured. I've been scrolling several screens back and still can't figure out the whole picture. My understanding is that nBlockSize accumulate the size of the block and nTxSize the size of the tx currently being considered for inclusion. Please correct me if that's wrong.

First, there is a problem because nBlockSize reserve space for the coinbase, and the sigops limit may end up being different in case the coinbase being used by the miner is smaller than what was reserved. I would argue that the size reserved for the coinbase shouldn't be accounted for when computing the sigops limit to be conservative.

zander replied:
CreateNewBlock is indeed huge. It is also a loop which makes it rather difficult to refactor in a more readable way.  I'm not arguing against a refactor, though ;)

jamoes replied:
Thank you for catching this! I have modified the code to subtract the coinbase reservation size when calculating the sigop limit in `CreateNewBlock`.

deadalnix commented on src/main.cpp:
```diff
@@ -3125,7 +3126,7 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
         nSigOps += GetLegacySigOpCount(tx);
     }
-    if (nSigOps > MAX_BLOCK_SIGOPS)
+    if (nSigOps > Policy::blockSigOpAcceptLimit(blockSize))
```
So this is using the old school sigops count, counting what is in the created outputs, but not what is in the spent outputs ? Is this what BU is doing ?

zander replied:
This is how Bitcoin counts non-p2sh addresses, yes.

Why do you think BU changed that, and if you feel that that is the case, that is something to bring up with BU.

jamoes replied:
According the the BUIP 40 document and my reading of the BU code, they are indeed using the `GetLegacySigOpCount` function for determining the sigop limits.

deadalnix reviewed (changes requested):
Overall it looks goods. Some details to improve and this can go.

zander commented:
I only have two changes that I think would be needed, first is the usage of 1e6 which I got convinced by deadalnix is not worth the risk.
And second is his point that its better to be safe in miner.cpp and use the size without the coinbase to calculate the sigops limit.

jamoes committed:
Updates from code review

Subtract potential size of coinbase tx when calculating sigop limit
in miner.cpp.

Add parens to blockSigOpAcceptLimit implementation for clarity.

Modify blockSigOpAcceptLimit unit tests to not use '1e6' notation, and
also choose better test values.

zquestz commented on src/miner.cpp:
```diff
@@ -251,7 +252,7 @@ CBlockTemplate* Mining::CreateNewBlock(const CChainParams& chainparams) const
                 continue;
 
-            const uint64_t maxSigOps = Policy::blockSigOpAcceptLimit(nBlockSize + nTxSize);
+            const uint64_t maxSigOps = Policy::blockSigOpAcceptLimit(nBlockSize + nTxSize - nCoinbaseReserveSize);
```
I am confused... nBlockSize == nCoinbaseReserveSize here... so this is essentially just Policy::blockSigOpAcceptLimit(nTxSize) which is not what I think you meant.

jamoes replied:
nBlockSize is incremented later on in the loop. Line 269 is: `nBlockSize += nTxSize`.

This function really needs to be refactored away from being one giant loop :)

zquestz replied:
Sure does! I missed that. Thanks for the clarification. Still getting used to the codebase. =)

zquestz commented on src/test/sigopcount_tests.cpp:
```diff
@@ -69,12 +69,16 @@ BOOST_AUTO_TEST_CASE(GetSigOpCount)
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(1700000), 40000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(1999999), 40000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(2000000), 40000);
+    BOOST_CHECK_EQUAL(Policy::blockSigOpAcceptLimit(2000001), 60000);
```
+1 Much cleaner and easier to read. =)
